File: app/src/main/java/com/arjanvlek/oxygenupdater/views/MessageDialog.java
package com.arjanvlek.oxygenupdater.views;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.view.KeyEvent;

/**
 * Usage: Title text, Message text, Positive button text, Negative button text.
 */
public class MessageDialog extends DialogFragment {
    private DialogListener dialogListener;

    private String title;
    private String message;
    private String positiveButtonText;
    private String negativeButtonText;
    private boolean closable;


    public interface DialogListener {
        void onDialogPositiveButtonClick(DialogFragment dialogFragment);
        void onDialogNegativeButtonClick(DialogFragment dialogFragment);
    }

    public MessageDialog setTitle(String title) {
        this.title = title;
        return this;
    }

    public MessageDialog setMessage(String message) {
        this.message = message;
        return this;
    }

    public MessageDialog setDialogListener(DialogListener listener) {
        dialogListener = listener;
        return this;
    }

    public MessageDialog setPositiveButtonText(String positiveButtonText) {
        this.positiveButtonText = positiveButtonText;
        return this;
    }

    public MessageDialog setNegativeButtonText(String negativeButtonText) {
        this.negativeButtonText = negativeButtonText;
        return this;
    }

    public MessageDialog setClosable(boolean closable) {
        this.closable = closable;
        return this;
    }


    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        builder.setTitle(title);
        builder.setMessage(message);

        if(negativeButtonText != null) {
            builder.setNegativeButton(negativeButtonText, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    if(dialogListener != null) {
                        dialogListener.onDialogNegativeButtonClick(MessageDialog.this);
                    }
                    dismiss();
                }
            });
        }
        if(positiveButtonText != null) {
            builder.setPositiveButton(positiveButtonText, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    if(dialogListener != null) {
                        dialogListener.onDialogPositiveButtonClick(MessageDialog.this);
                    }
                    dismiss();
                }
            });
        }

        if(!closable) {
            builder.setCancelable(false);
            builder.setOnKeyListener(new DialogInterface.OnKeyListener() {
                @Override
                public boolean onKey(DialogInterface dialogInterface, int i, KeyEvent keyEvent) {
                    if (i == KeyEvent.KEYCODE_BACK) {
                        getActivity().finish();
                        System.exit(0);
                    }
                    return true;
                }
            });
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                builder.setOnDismissListener(new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialogInterface) {
                        getActivity().finish();
                        System.exit(0);
                    }
                });
            }
        }
        return builder.create();

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if(!closable) {
            getActivity().finish();
            System.exit(0);
        }
    }
}

File: app/src/main/java/com/arjanvlek/oxygenupdater/dialogs/MessageDialog.kt
package com.arjanvlek.oxygenupdater.dialogs

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.os.Bundle
import android.os.Handler
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.DrawableRes
import androidx.core.view.isVisible
import com.arjanvlek.oxygenupdater.R
import com.arjanvlek.oxygenupdater.internal.KotlinCallback
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.button.MaterialButton
import kotlinx.android.synthetic.main.bottom_sheet_message.*
import kotlin.system.exitProcess

/**
 * Wrapper around [BottomSheetDialog]
 */
class MessageDialog(
    private val activity: Activity,
    private val title: String? = null,
    private val message: CharSequence? = null,
    private val positiveButtonText: String? = null,
    private val negativeButtonText: String? = null,
    @DrawableRes private val positiveButtonIcon: Int? = null,
    private val cancellable: Boolean = false,
    private val dialogListener: KotlinCallback<Int>? = null
) : BottomSheetDialog(activity) {

    @SuppressLint("InflateParams")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(LayoutInflater.from(activity).inflate(R.layout.bottom_sheet_message, null, false))
        setupViews()
    }

    private fun setupViews() {
        titleTextView.text = title
        messageTextView.text = message

        positiveButton.setup(positiveButtonText, View.OnClickListener {
            dismiss()
            dialogListener?.invoke(Dialog.BUTTON_POSITIVE)
        }, positiveButtonIcon)

        negativeButton.setup(negativeButtonText, View.OnClickListener {
            dismiss()
            dialogListener?.invoke(Dialog.BUTTON_NEGATIVE)
        })

        setupCancellableBehaviour()
    }

    private fun setupCancellableBehaviour() {
        setCancelable(cancellable)
        setCanceledOnTouchOutside(cancellable)

        if (!cancellable) {
            setOnKeyListener { _, keyCode, _ ->
                if (keyCode == KeyEvent.KEYCODE_BACK) {
                    dismiss()
                    exit()
                }
                true
            }

            setOnDismissListener { exit() }
        }
    }

    /**
     * This can be used to programmatically dismiss a dialog without triggering the dismiss listener,
     * which is set for non-cancellable dialogs
     */
    fun bypassListenerAndDismiss() {
        setOnKeyListener(null)
        setOnDismissListener(null)

        dismiss()

        // Reset cancellable behaviour, to
        setupCancellableBehaviour()
    }

    private fun MaterialButton.setup(string: String?, onClickListener: View.OnClickListener, @DrawableRes drawableResId: Int? = null) {
        string?.let {
            isVisible = true
            text = it
            setOnClickListener(onClickListener)
        }

        drawableResId?.let { icon = activity.getDrawable(it) }
    }

    private fun exit(activity: Activity? = this.activity) {
        if (activity != null) {
            activity.finishAffinity()
        } else {
            Handler().postDelayed({ exitProcess(0) }, 2000)
        }
    }
}
